Evaluate model on finer grid

olaurino opened the pull request:
## Release Note
Sherpa users can now define arbitrary grids, called evaluation spaces, on which to evaluate individual model components, both in 1D and 2D. This can be useful in a number of cases, for instance when it is desirable to evaluate models on a finer grid than the one defined by the data, or in convolution models where information outside of the data range can be used to reduce boundary effects or to inform the evaluation inside the data space. Also, when plotting individual 1D model components (`plot_source_component`), if a specific evaluation space was attached to the model then the plot will be of the model evaluated over that evaluation space, not the data space. Other plotting commands should be unaffected by this change.

## Notes
I based my work off what @DougBurke did in a branch on his fork, and then applied the following changes:

### `EvaluationSpace` class
First of all I created an `EvaluationSpace` class that takes care of the grid management, simplifying the `Regrid1D` implementations (all the tests pass with probably about half the code, which is also more readable and maintainable).

### Regridder without a grid
I didn't think a `Regrid1D` without a grid made much sense, so at first I removed such a possibility and some of the checks. For instance `Regrid1D` checked if the grid was none and just passed the original grid through. This seems overcomplicated. However, after going through more test cases, it seems like @DougBurke really wanted to have a way to say "this is a regrid model, but really it should simply pass through", and that one could even want to make the regrid mutable, i.e. to set the grid to `None` after instantiating it. I'd question the philosophy, but if that's the requirement, so be it. So rather than removing the tests and code I simply re-designed it to be cleaner and more explicit.

### `Regrid1D` and `Model`
I have a problem with redefining the semantics of `__call__` for `Regrid1D`. On the base class, __calc__ actually calculates the model, while for Regrid1D it just applies the regrid to a wrapped model. I tried making Regrid1D not extend Model, but then it can't be used to define a composite model. So either Regrid1D should indeed be a Model and __call__ should be used to calculate the model, or Regrid should not be a Model and the CompositeModel creation should be done differently, if at all.

Now, I really don't think `Regrid1D` should be a `Model`, because that has a number of issues (and the IDE agrees, there's a lot of warnings for a number of reasons), so that's what I did.

There is a test that exercises how we expect the regridded model to look like. It expected a composite model of two actual models, but that complicated things only for the sake of this choice, which seems arbitrary any way. So now the regridded model appears as a composite model with a name corresponding to the `Regrid1D` wrapper name + the name of the wrapped model, but the composition has only one part. Does that make any sense?

### Calculating the overlap between data space and eval space
Something that surprises me is that if we don't explicitly set the response to zero when there is no overlap the model actually evaluates to something that is not zero (I tried!). I am guessing it's because the extrapolation is just not going to zero? That's most likely true but I didn't dig further.
 
### Class names
I renamed `Regrid1D` as `ModelDomainRegridded1D`, which seems more explicit and self-documenting. And I renamed the `__call__` method as `apply_to`, which again seems more explicit.

### Functional testing
Unit tests were great for testing my refactoring (excellent work @DougBurke, thanks!) and I didn't need to change them unless I thought they were calling for bad design. I did remove some testing code that was dangling. We can bring them back, obviously.

### Test coverage
It's pretty good, but we are missing tests for the `ArithmeticFunctionModel` wrapper.

### API
I added a commit that addresses the high level API, i.e. how a user would use the UI to easily work with regridded models. In doing so I added a `regrid` method to the `Model` base class which takes care of instantiating the correct classes and returns an instance of the regridder wrapper. Unofruntately there is no way to know whether that makes sense for all of the `Model` subclasses. Maybe it's not important, but I wanted to make sure to note it.

### `MultigridSumModel`
While working on this PR I found this class defined, but it looks like it's never used in Sherpa. Does anybody know anything about it?

### `ArithmeticModel2D`
There is an issue in the current architecture that makes 1D and 2D models be interchangeable, which is however not true. In the scope of this PR, this makes it impossible (or rather it would require ugly workarounds) to override the `regrid` method for 2D models. I am introducing a new parent class for 2D models.
I made the cut in the most conservative way, i.e.:
  * models that declare ArithmeticModel as their superclass should be unaffected by the changes. However, they won't be regriddable. This includes user models, table models, and templates, at least for the time being.
  * models can declare ArithmeticModel1D or ArithmeticModel2D as the superclass. This will make them regriddable. I believe I changed the superclass of all the available models. Please let me know if 
I missed some or if I didn't mark them properly as 1D or 2D models.
  * given the new semantics, I renamed ArithmeticModel1D and its 2D counterpart as RegriddableModel1D and 2D, which makes more sense.

There are some models with a questionable class hierarchy, e.g. SigmaGauss2D is a Gauss2D, which is clearly not true. I didn't do anything about that though: if it ain't broken, don't fix it. However I did have to fix the questionable Const2D is a Const1D relation, so I extracted a common superclass Const(ArithmeticModel) and made Const1D and Const2D extend the proper superclass.

### Overlap conditions
In the original prototype by @DougBurke the overlap between spaces in the 1D case was pretty liberal, they would overlap as long as a portion of the interval was common to the evaluation spaces. For the 2D case, after a conversation with @juramaga we decided to stick to a more stringent condition that the spaces must have the same boundaries. Note that some requirements are not covered by neither set of conditions, as the dataset available to the model is only the one pre-processed by Sherpa and sent to the model for evaluation, which doesn't cover those cases where the model might want to peek outside of those boundaries. In that case the condition should rather be that the custom regrid region should *contain* the data region. **We really need to specify this stuff, although it might not make sense to do it in the context of this PR**, i.e. I'd rather have this PR merged as long as the general design seems reasonable and implements the basic use cases, and then extend or change the implementation when new information is available.

### TODOs and FIXMEs
We need to address those before we can merge this.

### EDITs
I removed the sections on 2D and the API improvements as despite the lack of feedback I moved forward with those. There is now a basic 2D implementation, and the API (as shown by the `test_regrid.py` tests) is more user-friendly, with much of the infrastructure hidden away.

olaurino commented:
> I have a problem with redefining the semantics of __call__ for Regrid1D. On the base class, calc actually calculates the model, while for Regrid1D it just applies the regrid to a wrapped model.

I was looking at the PSF code and I noticed that there is precedent for overloading the semantics of `__call__`, e.g. the `PSFModel` itself. I still think it's a bad idea (a really bad one, actually, as it violates OOP principles), and in the specific case of this PR, something we don't really need either.

juramaga commented:
Hi Omar,

I think we don't yet want to say that sherpa can evaluate the model on
*any* arbitrary grid. Or do we?

Rafael

On Fri, Jun 29, 2018 at 10:30 AM, Omar Laurino <notifications@github.com>
wrote:

> Release Note
>
> Sherpa users can now define arbitrary grids on which to evaluate
> individual models. This can be useful in a number of cases, for instance
> when it is desirable to evaluate models on a finer grid than the one
> defined by the data, or in convolution models where information outside of
> the data range can be used to reduce boundary effects or to inform the
> evaluation inside the data space.
> Notes
>
> *This is still WIP, but what's in here is ready to be reviewed*.
> I based my work off what @DougBurke <https://github.com/DougBurke> did in
> a branch on his fork, and then applied the following changes:
> EvaluationSpace class
>
> First of all I created an EvaluationSpace class that takes care of the
> grid management, simplifying the Regrid1D implementations (all the tests
> pass with probably about half the code, which is also more readable and
> maintainable).
> Regridder without a grid
>
> I didn't think a Regrid1D without a grid made much sense, so at first I
> removed such a possibility and some of the checks. For instance Regrid1D
> checked if the grid was none and just passed the original grid through.
> This seems overcomplicated. However, after going through more test cases,
> it seems like @DougBurke <https://github.com/DougBurke> really wanted to
> have a way to say "this is a regrid model, but really it should simply pass
> through", and that one could even want to make the regrid mutable, i.e. to
> set the grid to None after instantiating it. I'd question the philosophy,
> but if that's the requirement, so be it. So rather than removing the tests
> and code I simply re-designed it to be cleaner and more explicit.
> 2D
>
> I removed the 2D stuff since it's not exercised and just adds noise. The
> code is still versioned, so we can resuscitate it later. Should we include
> the 2D implementation before merging this? Not really sure, but leaning
> towards a yes, we need to, because one of the main applications for this is
> PSFs. But maybe we should just merge what we have and then revisit it for
> the 2D cases.
> Regrid1D and Model
>
> I have a problem with redefining the semantics of __call__ for Regrid1D.
> On the base class, *calc* actually calculates the model, while for
> Regrid1D it just applies the regrid to a wrapped model. I tried making
> Regrid1D not extend Model, but then it can't be used to define a composite
> model. So either Regrid1D should indeed be a Model and *call* should be
> used to calculate the model, or Regrid should not be a Model and the
> CompositeModel creation should be done differently, if at all.
>
> Now, I really don't think Regrid1D should be a Model, because that has a
> number of issues (and the IDE agrees, there's a lot of warnings for a
> number of reasons), so that's what I did.
>
> There is a test that exercises how we expect the regridded model to look
> like. It expected a composite model of two actual models, but that
> complicated things only for the sake of this choice, which seems arbitrary
> any way. So now the regridded model appears as a composite model with a
> name corresponding to the Regrid1D wrapper name + the name of the wrapped
> model, but the composition has only one part. Does that make any sense?
> Instantiating EvaluationSpace
>
> At this point users have to instantiate an EvaluationSpace class (see the
> tests). We can make that transparent by providing a convenience function
> that instantiate the correct class, and we could even hide that from the
> user by tweaking the Regrid1D initializer. I am just concerned that we
> complicate the Regrid classes by doing so we need to find an API that is
> simple for the user without complicating the design.
> Calculating the overlap between data space and eval space
>
> Something that surprises me is that if we don't explicitly set the
> response to zero when there is no overlap the model actually evaluates to
> something that is not zero (I tried!). I am guessing it's because the
> extrapolation is just not going to zero? That's most likely true but I
> didn't dig further.
> Class names
>
> I renamed Regrid1D as ModelDomainRegridded1D, which seems more explicit
> and self-documenting. And I renamed the __call__ method as apply_to,
> which again seems more explicit.
> Functional testing
>
> Unit tests were great for testing my refactoring (excellent work
> @DougBurke <https://github.com/DougBurke>, thanks!) and I didn't need to
> change them unless I thought they were calling for bad design. I did remove
> some testing code that was dangling, and the tests for the 2D stuff. We can
> bring them back, obviously. What it looks like we need now is to exercise
> the high level API we want to expose to users.
> Test coverage
>
> It's pretty good, but we are missing tests for the ArithmeticFunctionModel
> wrapper.
> API
>
> I added a commit that addresses the high level API, i.e. how a user would
> use the UI to easily work with regridded models. In doing so I added a
> regrid method to the Model base class which takes care of instantiating
> the correct classes and returns an instance of the regridder wrapper.
> Unofruntately there is no way to know whether that makes sense for all of
> the Model subclasses. Maybe it's not important, but I wanted to make sure
> to note it.
> MultigridSumModel
>
> While working on this PR I found this class defined, but it looks like
> it's never used in Sherpa. Does anybody know anything about it?
> ArithmeticModel2D
>
> There is an issue in the current architecture that makes 1D and 2D models
> be interchangeable, which is however not true. In the scope of this PR,
> this makes it impossible (or rather it would require ugly workarounds) to
> override the regrid method for 2D models. I am introducing a new parent
> class for 2D models, and using it for the model in the test. If this is
> approved and doesn't backfire, we should make sure all 2D models inherit
> from this one. I chose the narrowest context that seemed to make sense,
> i.e. ArithmeticModel, hoping that the change won't break something else
> downstream.
> Overlap conditions
>
> In the original prototype by @DougBurke <https://github.com/DougBurke>
> the overlap between spaces in the 1D case was pretty liberal, they would
> overlap as long as a portion of the interval was common to the evaluation
> spaces. For the 2D case, after a conversation with @juramaga
> <https://github.com/juramaga> we decided to stick to a more stringent
> condition that the spaces must have the same boundaries. Note that some
> requirements are not covered by neither set of conditions, as the dataset
> available to the model is only the one pre-processed by Sherpa and sent to
> the model for evaluation, which doesn't cover those cases where the model
> might want to peek outside of those boundaries. In that case the condition
> should rather be that the custom regrid region should *contain* the data
> region. *We really need to specify this stuff, although it might not make
> sense to do it in the context of this PR*, i.e. I'd rather have this PR
> merged as long as the general design seems reasonable and implements the
> basic use cases, and then extend or change the implementation when new
> information is available.
> TODOs and FIXMEs
>
> We need to address those before we can merge this.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/sherpa/sherpa/pull/469
> Commit Summary
>
>    - Experiment: add a regrid model to change model-evalution grid
>    - Initial support for 1D integrated models.
>    - Start on 2D cases.
>    - Fix for code that shouldn't have been committed
>    - some heavy refactoring
>    - make Regrid1D *not* a Model
>    - rename classes
>    - make doctest pass
>    - implement missing methods
>    - omit tests from coverage report
>    - use a pytest fixture for the setup
>    - improve API
>    - add 2d spaces, tests, and refactor
>    - extend start-end test
>    - implement first 2D API test, and implement what's missing
>    - properly implement 2d overlaps
>    - declare encoding
>
> File Changes
>
>    - *M* .coveragerc
>    <https://github.com/sherpa/sherpa/pull/469/files#diff-0> (4)
>    - *M* COPYRIGHT
>    <https://github.com/sherpa/sherpa/pull/469/files#diff-1> (4)
>    - *D* scripts/xspec.out
>    <https://github.com/sherpa/sherpa/pull/469/files#diff-2> (2136)
>    - *M* sherpa/astro/io/__init__.py
>    <https://github.com/sherpa/sherpa/pull/469/files#diff-3> (5)
>    - *M* sherpa/astro/utils/__init__.py
>    <https://github.com/sherpa/sherpa/pull/469/files#diff-4> (5)
>    - *M* sherpa/models/model.py
>    <https://github.com/sherpa/sherpa/pull/469/files#diff-5> (51)
>    - *A* sherpa/models/regrid.py
>    <https://github.com/sherpa/sherpa/pull/469/files#diff-6> (611)
>    - *A* sherpa/models/tests/test_regrid.py
>    <https://github.com/sherpa/sherpa/pull/469/files#diff-7> (343)
>    - *A* sherpa/models/tests/test_regrid_unit.py
>    <https://github.com/sherpa/sherpa/pull/469/files#diff-8> (899)
>    - *M* sherpa/utils/err.py
>    <https://github.com/sherpa/sherpa/pull/469/files#diff-9> (2)
>
> Patch Links:
>
>    - https://github.com/sherpa/sherpa/pull/469.patch
>    - https://github.com/sherpa/sherpa/pull/469.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sherpa/sherpa/pull/469>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFP5A8wy5geY6wQtgk6V0-4sHlVZGY96ks5uBjofgaJpZM4Ur4t2>
> .
>



-- 
------------------------------------------------------
J. Rafael Martinez-Galarza
Harvard-Smithsonian Center for Astrophysics
60 Garden Street - MS 66
Cambridge, MA 02138
U. S. A.
+1 617 495 7027
jmartine@cfa.harvard.edu
------------------------------------------------------

olaurino commented:
@juramaga if you refer to the contents of the release note, that's what we'll include in the release notes when the PR is merged. It can be rephrased to match changes we decide to make, but it should describe what Sherpa will be able to do after the work on this PR is done.

olaurino commented:
@anetasie @DougBurke could you please review this?

olaurino commented:
@anetasie I am not sure I understand your request.

> How can I validate that the model is evaluated on the original grid, when it is being fit to the data.

Not sure I understand. When you set the model to be evaluated on a different grid, the grid being used for the fit is the arbitrary grid set by the user, not the data one. The result of the evaluation is then rebinned back to the data-defined grid.

> When I display the results the model is displayed on the data grid. Is there a way to access the model on the grid?

If you mean plotting no, you had set a lower priority on plotting, so that's not done yet.

olaurino commented:
You can evaluate the model like every other model, by calling it with an array as the argument. The purpose of the regridding is to make that happen during the fit, when the user doesn't currently have control over the evaluation grid. Does that answer your question?

olaurino commented:
After today's meeting @anetasie requested that in the 1D case the arbitrary grid and the data grid need to be concatenated, so the data grid is always included in the model evaluation.

olaurino commented:
All right, I think I have a suitable fix to the functional regressions introduced into the tests by implementing the new requirement. However, there are a number of unit tests that are failing because of the changed requirement and I'll have to address those. They should be easier to address, but it's a number of them. I'll do that next week.

olaurino commented:
I published the new code. There are two unit tests that I still expect to fail. They have to do with integrated models. They test that the models evaluate to the same values on different grids. I don't think the models should be evaluating to the same numbers, though, as the grids have different binnings and the models are integrated. The values I get make sense to me, but maybe I am missing something. @anetasie can we have a chat sooner rather than later?

I think the tests were previously passing because the models had different parameter values and they might have evaluated to _close enough_ values. I had to change the model parameters to have better control over the expected results for the unit tests.

olaurino commented:
I updated the interpolation/rebinning logic at @anetasie and I discussed this morning, and things seem to finally come together. If Travis agrees I'll start refactoring the code.

Note that I used a `rebin` function that we already had implemented, so that's good. I'll also double check that the assumption we made, i.e. that when the dataspace is binned and the integrate flag is False Sherpa picks the midpoint of the bins to evaluate the model point-wise.

olaurino commented:
Some tests were failing on python 2.7 because of an old-style division rounding error. That should be fixed now.

olaurino commented:
While reviewing the sherpa code that deals with integrated models I realized that when the integrate flag is False and two x arrays are passed sherpa uses the first array, not the midpoint. So in the regrid code I decided to pass the midpoint grid (@anetasie seems to agree).

There is still a test failing for reasons completely unrelated to this PR. The problem seems to be with xvfb not working in a specific case, so the plot test does not complete. I'll see on Monday whether I can fix it or I'll just let the test be skipped is a display is not available.

olaurino commented:
Rebasing to resolve conflicts.

olaurino commented:
I expect the latest commit to work in Travis. @anetasie if you could give it another quick review before I merge it that would be helpful. Thanks!

Especially the new plotting support for arbitrary grids.

olaurino commented:
With perfect timing, there is a completely unrelated problem with the ds9 tests. The older ds9 version link is not working, and when updating to the new version (a217363) the tests error out.

olaurino commented:
Should be fine now.

anetasie commented:
@olaurino I tried and the following fails with the error. Note that I need to do `from sherpa.astro.ui import * ` . When I tried `import sherpa.astro.ui as ui`  ui.dataspace1d(1,100,2,dstype=Data1D) gives me the error:
```
In [1]: import sherpa.astro.ui as ui
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available

In [2]: ui.dataspace1d(1,100,2,dstype=Data1D)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-2-1e25f42aa166> in <module>()
----> 1 ui.dataspace1d(1,100,2,dstype=Data1D)

NameError: name 'Data1D' is not defined
````

So I just did the following:

```
from sherpa.astro.ui import *
ui.dataspace1d(1,100,2,dstype=Data1D)
dataspace1d(1,100,2,dstype=Data1D)
set_model(gauss1d.g0)
set_par(g0.fwhm,20)
set_par(g0.pos, 50)
set_par(g0.ampl, 100.)
fake()
import numpy as np
import matplotlib
%matplotlib
plot_data()
grid=np.arange(100)
one = np.ones(100)
grid = grid+one
set_model(gauss1d.g0)
plot_model()
print get_model_plot()
r0=g0.regrid(grid)
print r0
set_model(r0)
plot_model()  # this gives an error
grid
```
Here is the error:

```
plot_model()
sherpa/data.py:606: UserWarning: evaluation space does not contain the requested space. Sherpa will join the two spaces.
  warnings.warn("evaluation space does not contain the requested space. Sherpa will join the two spaces.")
sherpa/models/regrid.py:400: UserWarning: evaluation space does not contain the requested space. Sherpa will join the two spaces.
  warnings.warn("evaluation space does not contain the requested space. Sherpa will join the two spaces.")

```
and with fit()

```
....
/Users/aneta/git/sherpa/sherpa/stats/__init__.pyc in calc_stat(self, data, model)
    504                           fitdata[1], fitdata[2],
    505                           None,  # TODO: weights
--> 506                           truncation_value)
    507 
    508     def calc_chisqr(self, data, model):

TypeError: statistic input array sizes do not match

```

olaurino commented:
@anetasie up to the fit error it is working as expected. We might review the way the warnings are displayed, but other than that it's working as expected.

You are creating a 1D data space with bin size 2, from 1 to 101. Then you create an evaluation space with bin size 1, from 1 to 100. Since the data space is not contained in the evaluation space, the two spaces are joined, as per your requirements, hence the warning.

The function you asked me to support arbitrary grids is `plot_model_component`. You can see the code at work by creating a data space with bin size 10, and then use `plot_model_component` with g0 first and then r0, overplotting. That will show clearly they are being evaluated on two different grids with different bin sizes.

I will look into the fit error.

olaurino commented:
I believe I have a fix. I'll test it some more and clean it up and I'll push it tomorrow.

anetasie commented:
@olaurino I just run the updated code (4.10.0-59-g0d4e422) and things seem to work ok for a simple test. I'm wondering about changing 'regrid' to 'evaluate' which is closer to what is happening to the model...

anetasie commented:
@olaurino I'm looking at the integrated model. The fit parameters look ok, but the plotted model does not plots the integrated over the bin values, so it is a factor 2 below the data points in ```plot_fit()```

```
dataspace1d(1,100,2,dstype=Data1DInt)
set_model(gauss1d.g1)
set_par(g1.fwhm,20)
set_par(g1.pos, 50)
set_par(g1.ampl,100.)
fake()
print(get_data())
fit()

# new evaluation grid:
grid_hi = np.linspace(2,101,100)
grid_lo = np.linspace(1,100,100)
r1 = g1.regrid(grid_lo,grid_hi)
set_model(r1)
fit()
plot_model()
```

olaurino commented:
I'll look into it. We have tests that prove the plot commands do not error out, but we don't have tests that prove the plot results are correct (other than the one I added for the arbitrary grid), so it's impossible to get this kind of regressions from the automated tests. Let me see what I can do to improve things. Did you test that the `plot_source_component` provides the right plot?

olaurino commented:
The plot values are correct. The problem is that `plot_model` is plotting the model as evaluated on the arbitrary grid you set, and since that grid has half the bin size of the data grid, the values are 1/2 the values evaluated over the data space.

@anetasie confirmed that `plot_model` should plot the model as evaluated on the data space, not the evaluation space.

olaurino commented:
The issue with `plot_model` should now be fixed. I changed the code in such a way that only `plot_source_model` should behave differently. Given the nature of the code involved and the lack of tests I can't guarantee I didn't break anything, though :)

anetasie commented:
@olaurino I can confirm that `plot_model`, `plot_model_component' and `plot_fit` give in correct plot for data type `Data1DInt`. However,  there are still issues with `plot_source` which does not use the evaluation grid.

Here is the evaluation grid from the previous test:
```
# new evaluation grid:
grid_hi = np.linspace(2,101,100)
grid_lo = np.linspace(1,100,100)
r1 = g1.regrid(grid_lo,grid_hi)
```

Here is the plot_source and plot_source_component grid, which has a bin size of 2 instead of 1.

```
In [20]: print(get_source_component_plot(r1))
x      = [   1.5,   2.5,   3.5,   4.5,   5.5,   6.5,   7.5,   8.5,   9.5,  10.5,
   11.5,  12.5,  13.5,  14.5,  15.5,  16.5,  17.5,  18.5,  19.5,  20.5,
   21.5,  22.5,  23.5,  24.5,  25.5,  26.5,  27.5,  28.5,  29.5,  30.5,
   31.5,  32.5,  33.5,  34.5,  35.5,  36.5,  37.5,  38.5,  39.5,  40.5,
   41.5,  42.5,  43.5,  44.5,  45.5,  46.5,  47.5,  48.5,  49.5,  50.5,
   51.5,  52.5,  53.5,  54.5,  55.5,  56.5,  57.5,  58.5,  59.5,  60.5,
   61.5,  62.5,  63.5,  64.5,  65.5,  66.5,  67.5,  68.5,  69.5,  70.5,
   71.5,  72.5,  73.5,  74.5,  75.5,  76.5,  77.5,  78.5,  79.5,  80.5,
   81.5,  82.5,  83.5,  84.5,  85.5,  86.5,  87.5,  88.5,  89.5,  90.5,
   91.5,  92.5,  93.5,  94.5,  95.5,  96.5,  97.5,  98.5,  99.5, 100.5]
```

olaurino commented:
@anetasie that looks like the proper grid to plot for the space you are setting. What am I missing?

anetasie commented:
@olaurino Yes, the grid looks ok, but there is still something which is not correct. For the integrated data set the total flux is not being preserved and the total number of source counts depends on the bin size.

```
from sherpa.astro.ui import *
import numpy as np
import matplotlib
%matplotlib

dataspace1d(1,100,2,dstype=Data1DInt)
set_model(box1d.b1)
b1.xlo=10
b1.xlow=10
b1.xhi=80
b1.ampl.max=100
show_model()
b1.ampl=10
fake()

plot_data()
plot_model()
plot_source()

fit()
plot_fit()

grid_hi = np.linspace(2,101,200)
grid_lo = np.linspace(1,100,200)
rb1 = b1.regrid(grid_lo,grid_hi)
set_model(rb1)
plot_source_component(rb1)
show_model()
fit()
plot_fit()
np.sum(get_source_component_plot(rb1).y)
np.sum(get_source_component_plot(b1).y)
calc_data_sum()
plot_source_component(rb1,overplot=True)

print get_source_component_plot(rb1).x
grid_hi = np.linspace(2,101,100)
grid_lo = np.linspace(1,100,100)
rb1 = b1.regrid(grid_lo,grid_hi)
print get_source_component_plot(rb1).x
np.sum(get_source_component_plot(rb1).y)
```

olaurino commented:
I am looking into it. I am trying to understand which numbers are not what you would expect. It would help if you could use `assert` to show what you expect.

anetasie commented:
I was just chatting with @DougBurke about the plotting. Perhaps the issue is related to the units of the plot. We plot 'counts/bin' in PHA files, so the division or multiplication by bin is missing. What I expect to see is the source component overplotted on the data having the high resolution grid and overall and overall normalization as the data, so the bin size difference between the two matter.

```
plot_fit()
plot_source_component(rb1,overplot=True)

```

olaurino commented:
@anetasie so what is the request in terms of `plot_source_component`?

Going back to the question whether the flux is conserved or not, I am getting some strange results using the Box1D model. My suspicion is that something is wrong when the model, evaluated on the evaluation space, is then rebinned to the data space. I don't get the same issues with the constant model. There might be something wrong with the rebinning code, the model, or something else, but I am looking into it.

I get the expected results when the model is constant, which should be close enough to the Box1D case, but apparently it's not.

olaurino commented:
I am almost certain there is a bug in the `Box1D` model. It's the only one that behaves differently from the others and doesn't give the expected results. In fact, if you look at its definition, you'll see the integrated version does not calculate the appropriate value. Compare:

https://github.com/sherpa/sherpa/blob/c1a48243bc8fe53123098644710bbf3c2f7af6a8/sherpa/include/sherpa/models.hh#L61

to:

https://github.com/sherpa/sherpa/blob/c1a48243bc8fe53123098644710bbf3c2f7af6a8/sherpa/include/sherpa/models.hh#L93

olaurino commented:
If in the definition of the box1d model I remove the division for `xhi - xlo` then I get the expected results. I'll wait for @anetasie and @DougBurke to confirm my conclusion, then I'll be happy to include the fix to `Box1D` in this PR.

I am still not sure if there is an extra requirement for `plot_source_component`. @anetasie could you please clarify?

anetasie commented:
@olaurino it does look like a bug in ```Box1D``` model. ```plot_model``` look wrong with the amplitude falling down instead of being constant within the limits.

I'm still trying to figure out what we should do for `plot_source_component`. The current `plot_source_component` does not show the expected normalization for the integrated data.

olaurino commented:
Rebasing to fix conflicts. Was bc47fe6.

juramaga commented:
I am trying to test the 2D fitting within a Jupyter notebook. I have some test data. I ran into a problem pretty early on, when I was trying to read the image data:

# Let us first load the image data into Sherpa
from sherpa.astro import ui
import numpy as np
ui.load_image('data/acisf02594_001N999_r0008s_ctsimg3.fits')

This killed the kernel. Any idea why this could be happening?

olaurino commented:
@juramaga that's unlikely to be related to this PR and most likely be a local issue. I'll follow up with you directly.

juramaga commented:
For the 2D case, do we want the evaluation space into which we will regrid to be a meshgrid? What does the regrid attribute expect as an input in this case?

I have some test data in which the pixel size in the PSF is twice as small as the pixel size in the data space. So, if the pixels size is =1 for the PSF space, and the PSF image has dimensions 42x42 (data has dimensions 21x21, but they cover the same area), I was thinking of doing something like:

```

# Create meshgrid with the points where I want the model to be evaluated
x = np.arange(42)
y = np.arange(42)
xv, yv = np.meshgrid(x, y)

# Regrid model to the desired evaluation space
model.regrid(xv,yv)

fit()

```

But I am not sure if that makes sense.

juramaga commented:
So far I have been unsuccessful at running the 2D rebinning/fitting in some test data. I am trying to fit a Gaussian model to an s band image. The image has dimensions 52x48 image pixels, but in physical pixels the dimensions are 104x96. All I am doing is loading the data, setting a finer grid on which I want to evaluate the model, then define my model, regrid it to the arbitrary grid, and fitting (code is below). I am aware that the evaluation and data grid should completely overlap, but I do not know in which units. I also copy below the error I get.

Code:

```

# Let us first load the image data into Sherpa
from sherpa.utils.testing import requires_data, requires_fits
from sherpa.astro import ui
ui.load_image('data/acisf02594_001N999_r0008s_ctsimg3.fits')

# Now I define a finer grid, with twice as many pixels. Here is where I do not know if
# I should use image or physical pixels. I have tried both, and failed.
x = linspace(0,48,48*2)
y = linspace(0,52,52*2)

# We now set a Gaussian source model 
ui.set_source(ui.gauss2d.mdl)
mdl.ellip.thaw()
mdl.theta.thaw()

# Now let's regrid and set the source to the regrid model.
rb1 = mdl.regrid(x,y)
ui.set_source(rb1)

# Here we set the statistic to the Cash statisitc, and the optimization method to Simplex.
ui.set_stat('cstat')
ui.set_method('simplex')

# Now fit
ui.guess(rb1)
#ui.fit()

```

And this is the error I get:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-8-95011a46f8f7> in <module>()
----> 1 ui.fit()

<string> in fit(id, *otherids, **kwargs)

~/software/sherpa/sherpa/astro/ui/utils.py in fit(self, id, *otherids, **kwargs)
   9372         """
   9373         kwargs['bkg_only'] = False
-> 9374         self._fit(id, *otherids, **kwargs)
   9375 
   9376     def fit_bkg(self, id=None, *otherids, **kwargs):

~/software/sherpa/sherpa/astro/ui/utils.py in _fit(self, id, *otherids, **kwargs)
   9476                     i).mask & numpy.isfinite(self.get_data(i).get_x())
   9477 
-> 9478         res = f.fit(**kwargs)
   9479         res.datasets = ids
   9480         self._fit_results = res

~/software/sherpa/sherpa/fit.py in run(fit, *args, **kwargs)
     51     def run(fit, *args, **kwargs):
     52         fit.model.startup()
---> 53         result = func(fit, *args, **kwargs)
     54         fit.model.teardown()
     55         return result

~/software/sherpa/sherpa/fit.py in fit(self, outfile, clobber)
   1262             raise FitErr('binhas0')
   1263 
-> 1264         init_stat = self.calc_stat()
   1265         # output = self.method.fit ...
   1266         output = self._iterfit.fit(self._iterfit._get_callback(outfile, clobber),

~/software/sherpa/sherpa/fit.py in calc_stat(self)
   1140         """
   1141 
-> 1142         return self._calc_stat()[0]
   1143 
   1144     def calc_chisqr(self):

~/software/sherpa/sherpa/fit.py in _calc_stat(self)
   1121         # TODO: is there anything missing here that
   1122         #       self._iterfit.get_extra_args calculates?
-> 1123         return self.stat.calc_stat(self.data, self.model)
   1124 
   1125     def calc_stat(self):

~/software/sherpa/sherpa/stats/__init__.py in calc_stat(self, data, model)
    288 
    289     def calc_stat(self, data, model):
--> 290         fitdata, modeldata = self._get_fit_model_data(data, model)
    291 
    292         return self._calc(fitdata[0], modeldata, None,

~/software/sherpa/sherpa/stats/__init__.py in _get_fit_model_data(self, data, model)
    197         data, model = self._validate_inputs(data, model)
    198         fitdata = data.to_fit(staterrfunc=self.calc_staterror)
--> 199         modeldata = data.eval_model_to_fit(model)
    200 
    201         return fitdata, modeldata

~/software/sherpa/sherpa/data.py in eval_model_to_fit(self, modelfuncs)
    516 
    517         for func, data in izip(modelfuncs, self.datasets):
--> 518             total_model.append(data.eval_model_to_fit(func))
    519 
    520         return numpy.concatenate(total_model)

~/software/sherpa/sherpa/data.py in eval_model_to_fit(self, modelfunc)
    188 
    189     def eval_model_to_fit(self, modelfunc):
--> 190         return modelfunc(*self.get_indep(filter=True, model=modelfunc))
    191 
    192     #

~/software/sherpa/sherpa/models/model.py in __call__(self, *args, **kwargs)
    267         if (len(args) == 0 and len(kwargs) == 0):
    268             return self
--> 269         return self.calc([p.val for p in self.pars], *args, **kwargs)
    270 
    271     def _get_thawed_pars(self):

~/software/sherpa/sherpa/models/model.py in calc(self, p, *args, **kwargs)
    746 
    747     def calc(self, p, *args, **kwargs):
--> 748         return self.wrapper.calc(p, self.model.calc, *args, **kwargs)
    749 
    750     def get_center(self):

~/software/sherpa/sherpa/models/regrid.py in calc(self, pars, modelfunc, *args, **kwargs)
    360         requested_eval_space = self._make_and_validate_grid(args)
    361 
--> 362         return self._evaluate(requested_eval_space, pars, modelfunc, **kwargs)
    363 
    364     def _make_and_validate_grid(self, args_array):

~/software/sherpa/sherpa/models/regrid.py in _evaluate(self, data_space, pars, modelfunc, **kwargs)
    413                 # This should be the most common case
    414                 y = modelfunc(pars, evaluation_space.grid[0], evaluation_space.grid[1],
--> 415                               **kwargs)
    416                 return rebin(y,
    417                              evaluation_space.grid[0], evaluation_space.grid[1],

~/software/sherpa/sherpa/models/basic.py in calc(self, *args, **kwargs)
   1480     def calc(self, *args, **kwargs):
   1481         kwargs['integrate'] = bool_cast(self.integrate)
-> 1482         return _modelfcts.gauss2d(*args, **kwargs)
   1483 
   1484 

TypeError: 2D model evaluation input array sizes do not match

```

olaurino commented:
@juramaga thanks so much for testing this. I hadn't realized I hadn't completed the 2D implementation. See this bit in my PR description:

>ArithmeticModel2D
There is an issue in the current architecture that makes 1D and 2D models be interchangeable, which is however not true. In the scope of this PR, this makes it impossible (or rather it would require ugly workarounds) to override the regrid method for 2D models. I am introducing a new parent class for 2D models, and using it for the model in the test. **If this is approved and doesn't backfire, we should make sure all 2D models inherit from this one**. I chose the narrowest context that seemed to make sense, i.e. ArithmeticModel, hoping that the change won't break something else downstream.

So the 2D code only worked for my tests because I was using the 2D superclass, but the Sherpa 2D models weren't updated.

I'll see how bad it is to update all the models at this point. As Sherpa currently doesn't differentiate between 1D and 2D models, it might be a big issue to only have the 1D code work. Maybe we could caveat?

I was expecting to get feedback in a few days, I got it after more than 2 months and I forgot. Note to self: use ToDo lists in the PR description.

@juramaga thanks again for testing!

olaurino commented:
I assumed people are fine with differentiating 1D and 2D models. I made the cut in the most conservative way, i.e.:

  * models that declare `ArithmeticModel` as their superclass should be unaffected by the changes. However, they won't be regriddable. This includes user models, table models, and templates, at least for the time being.
  * models can declare `ArithmeticModel1D` or `ArithmeticModel2D` as the superclass. This will make them regriddable. I believe I changed the superclass of all the available models. **Please let me know if I missed some or if I didn't mark them properly as 1D or 2D models**.
  * given the new semantics, I renamed `ArithmeticModel1D` and its 2D counterpart as `RegriddableModel1D` and 2D, which makes more sense.

There are some models with a questionable class hierarchy, e.g. `SigmaGauss2D` *is a* `Gauss2D`, which is clearly not true. I didn't do anything about that though: if it ain't broken, don't fix it. However I did have to fix the questionable `Const2D` *is a* `Const1D` relation, so I extracted a common superclass `Const(ArithmeticModel)` and made `Const1D` and `Const2D` extend the proper superclass.

There was also an issue in the way I called the function responsible for rebinning in 2D, which meant that only images with the same number of pixels on `x` and `y` were properly treated. This is, ultimately, what was making @juramaga's tests fail, since his real-world image did not have the same number of pixels on the two axes.

I am going to push a new commit soon. If I don't get any feedback I'll merge this PR anyway, but **people should review it nevertheless**, as if there are issues we need to fix them ASAP.

anetasie commented:
@olaurino yes, 1D and 2D models can be differentiated. We are already doing it, so that's not difficult for a user and in fact less confusing. I'll test the new code when it is ready.

olaurino commented:
By the way, we had an old bug in our legacy bug tracking system:  

>no complaint about combining a 1D and 2D model

I added it to the new bug tracking system, since we can now fix it (except that there are still models that are not 1D or 2D yet, so for those models I wouldn't know what to do. In any case we can try addressing that issue now that there is a difference.

anetasie commented:
I am generating a regression test in Sherpa/CIAOT and I see that both model on the data grid and on a smaller grid are being plotted with the same bin sizes. It looks like there is still an issue in the way the model components are plotted.

'''
dataspace1d(1,100,2,dstype=Data1D)
set_model(gauss1d.g0)
set_par(g0.fwhm,20)
set_par(g0.pos, 50)
set_par(g0.ampl, 100.)
fake()
print(get_data_plot().x)   # this is the data bins

grid=np.arange(100)
one = np.ones(100)
grid = grid+one

set_model(gauss1d.g0)
plot_model()
print(get_model_plot().x) # this is the data bins

# Now  define a model g0 on a grid

r0=g0.regrid(grid)
print(r0)
set_model(r0)
plot_model()  # this gives a warning and plots the model on data bins -ok

fit() # this gives a lot of warnings, but returns the fit - ok.
plot_model_component(g0)
plot_model_component(r0,overplot=True)
'''

'''
sherpa In [29]: get_model_component_plot(r0).x
/proj/xena/ciaot_install/LinuxP3.181101/ciao-4.11/lib/python3.5/site-packages/sherpa-ciao4.11-py3.5-linux-x86_64.egg/sherpa/models/regrid.py:400: UserWarning: evaluation space does not contain the requested space. Sherpa will join the two spaces.
  warnings.warn("evaluation space does not contain the requested space. Sherpa will join the two spaces.")
Out[29]: 
array([   1.,    3.,    5.,    7.,    9.,   11.,   13.,   15.,   17.,
         19.,   21.,   23.,   25.,   27.,   29.,   31.,   33.,   35.,
         37.,   39.,   41.,   43.,   45.,   47.,   49.,   51.,   53.,
         55.,   57.,   59.,   61.,   63.,   65.,   67.,   69.,   71.,
         73.,   75.,   77.,   79.,   81.,   83.,   85.,   87.,   89.,
         91.,   93.,   95.,   97.,   99.,  101.])

sherpa In [30]: get_model_component_plot(g0).x
Out[30]: 
array([   1.,    3.,    5.,    7.,    9.,   11.,   13.,   15.,   17.,
         19.,   21.,   23.,   25.,   27.,   29.,   31.,   33.,   35.,
         37.,   39.,   41.,   43.,   45.,   47.,   49.,   51.,   53.,
         55.,   57.,   59.,   61.,   63.,   65.,   67.,   69.,   71.,
         73.,   75.,   77.,   79.,   81.,   83.,   85.,   87.,   89.,
         91.,   93.,   95.,   97.,   99.,  101.])

'''

olaurino commented:
@anetasie as per requirements only `plot_source_component` is affected by this change. I tried your script and when I use `plot_source_component` I get the expected results.

anetasie commented:
Yes. I realized that this is the case. We will need to revisit the plotting for the arbitrary grid.

I just tried to use integrated data set with box model. The model plot is incorrect, as it shows the 'saw pattern'. Here is the script:

```
dataspace1d(1,100,2,dstype=Data1DInt)
set_model(box1d.b1)
b1.xlow=10
b1.xhi=80
b1.ampl.max=100
show_model()
b1.ampl=10
fake()
fit()
plot_fit()  # this look ok

# setting up the grid
grid_hi = np.linspace(2,101,600)
grid_lo = np.linspace(1,100,600)
rb1 = b1.regrid(grid_lo,grid_hi)
set_model(rb1)
fit()
plot_fit()  # this one shows the issue with the model line
plot_model()
plot_source_component(rb1,overplot=True) 
```
